### PR TITLE
Allow rate limiting before auth filters

### PR DESCRIPTION
**Pull request by GrahamGoudeau**
Allow running the rate limit filter before the auth filters. *NOTE* that, when this new setting is used, it will result in the loss of the extauth-aware rate limiting feature (ie, different rate limits for authenticated vs unauthenticated users)

ext auth through basic auth:
```
(⎈ |minikube:gloo-system)~/go/src/github.com/solo-io/solo-projects > kubectl edit authconfigs.enterprise.gloo.solo.io -oyaml
apiVersion: enterprise.gloo.solo.io/v1
kind: AuthConfig
metadata:
...
  name: basic-auth
  namespace: gloo-system
...
spec:
  configs:
  - basicAuth:
      apr:
        users:
          user:
            hashedPassword: 6EeY/AURxHUpm4EuRi/Ck/
            salt: UPYWMjsk
```
settings:
```
(⎈ |minikube:gloo-system)~/go/src/github.com/solo-io/solo-projects > kubectl get settings -oyaml
apiVersion: v1
items:
- apiVersion: gloo.solo.io/v1
  kind: Settings
  metadata:
...
    name: default
    namespace: gloo-system
...
  spec:
    discoveryNamespace: gloo-system
    extensions:
      configs:
        extauth:
          extauthzServerRef:
            name: extauth
            namespace: gloo-system
        rate-limit:
          deny_on_fail: true
          rate_limit_before_auth: true
          ratelimit_server_ref:
            name: rate-limit
            namespace: gloo-system
...
    refreshRate: 6s
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
virtual service:
```
(⎈ |minikube:gloo-system)~/go/src/github.com/solo-io/solo-projects > kubectl get vs -oyaml
apiVersion: v1
items:
- apiVersion: gateway.solo.io/v1
  kind: VirtualService
  metadata:
...
    name: default
    namespace: gloo-system
...
  spec:
    virtualHost:
      domains:
      - '*'
      routes:
      - matchers:
        - exact: /sample-route-1
        routeAction:
          single:
            upstream:
              name: default-petstore-8080
              namespace: gloo-system
        routePlugins:
          prefixRewrite: /api/pets
      virtualHostPlugins:
        extauth:
          configRef:
            name: basic-auth
            namespace: gloo-system
        extensions:
          configs:
            rate-limit:
              anonymous_limits:
                requests_per_unit: 1
                unit: SECOND
...
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1445

### CODE REVIEW

File: changelog/v0.21.0/filter-ordering.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+changelog:
+  - type: NEW_FEATURE
+    description: Allow running the rate limit filter before the auth filters
```
Comment (marcogschmidt) [inline changelog/v0.21.0/filter-ordering.yaml]: Can you add a quick disclaimer about the fact that enabling this configuration will result in loss of the "extauth-aware" rate limiting feature? (also add it to the PR description)

File: changelog/v0.21.0/filter-ordering.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+changelog:
+  - type: NEW_FEATURE
+    description: |
```
Comment (kdorosh) [inline changelog/v0.21.0/filter-ordering.yaml]: nit: formatting
```suggestion
    description: >
```

File: changelog/v0.21.0/filter-ordering.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+      Allow running the rate limit filter before the auth filters. *NOTE* that this will result
+      in the loss of the extauth-aware rate limiting feature.
```
Comment (kdorosh) [inline changelog/v0.21.0/filter-ordering.yaml]: in parens, would be nice to clarify what the "extauth-aware rate limiting feature" is.

"Basic Rate Limiting", "Gloo Rate Limiting", and "Ingress Rate Limiting" are all terms we have thrown around in the code/docs for this feature, although I prefer the term "basic rate limiting myself"

Comment (kdorosh) [reply changelog/v0.21.0/filter-ordering.yaml]: also, we should be explicitly clear that the loss of the extauth-aware rate limiting feature only happens when you run ratelimit filters before auth filters. As written, this changelog might be interpreted as a breaking change

File: projects/gloo/pkg/plugins/ratelimit/plugin.go [hunk 4]
```diff
@@ -44,6 +49,7 @@ func (p *Plugin) Init(params plugins.InitParams) error {
 		p.upstreamRef = rlServer.RatelimitServerRef
 		p.timeout = rlServer.RequestTimeout
 		p.denyOnFail = rlServer.DenyOnFail
+		p.rateLimitBeforeAuth = params.Settings.GetRatelimitServer().RateLimitBeforeAuth
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/ratelimit/plugin.go]: if ratelimit isn't configured at all, then this plugin will panic. the plugin should be able to initialize properly and immediately return during the `Process*` functions if the config is incomplete
```suggestion
		p.rateLimitBeforeAuth = params.Settings.GetRatelimitServer().GetRateLimitBeforeAuth()
```

File: projects/gloo/pkg/plugins/ratelimit/plugin_test.go [hunk 5]
```diff
@@ -89,6 +95,50 @@ var _ = Describe("Plugin", func() {
...
+					RequestTimeout: &timeout,
+				},
+			}
+			rlPlugin.Init(initParams)
+		})
+
+		It("should be ordered before ext auth", func() {
+			filters, err := rlPlugin.HttpFilters(params, nil)
+			Expect(err).NotTo(HaveOccurred(), "Should be able to build rate limit filters")
+			Expect(len(filters)).To(Equal(1), "Should only have created one custom filter")
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/ratelimit/plugin_test.go]: more idiomatic to do
```suggestion
			Expect(filters).To(HaveLen(1), "Should only have created one custom filter")
```

**Pull request merged**